hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/ContentSummaryProcessor.java

```diff
@@ -0,0 +1,109 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.azurebfs.services;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.azurebfs.AzureBlobFileSystemStore;
+import org.apache.hadoop.fs.azurebfs.utils.ABFSContentSummary;
+
+public class ContentSummaryProcessor {
+  private final AtomicLong fileCount = new AtomicLong(0L);
+  private final AtomicLong directoryCount = new AtomicLong(0L);
+  private final AtomicLong totalBytes = new AtomicLong(0L);
+  private final LinkedBlockingQueue<FileStatus> queue = new LinkedBlockingQueue<>();
+  private final Logger LOG =
+      LoggerFactory.getLogger(ContentSummaryProcessor.class);
+  private final AzureBlobFileSystemStore abfsStore;
+  private static final int NUM_THREADS = 16;
+  private final AtomicInteger NUM_TASKS = new AtomicInteger(0);
+  ExecutorService executorService = new ThreadPoolExecutor(1, NUM_THREADS,
+      5, TimeUnit.SECONDS, new SynchronousQueue<>());
+
+  public ContentSummaryProcessor(AzureBlobFileSystemStore abfsStore) {
+    this.abfsStore = abfsStore;
+  }
+
+  public ABFSContentSummary getContentSummary(Path path)
+      throws IOException, InterruptedException {
+    processDirectoryTree(path);
+
+    while(!queue.isEmpty() || NUM_TASKS.get() > 0) {
+      Thread.sleep(10);
+    }
+    executorService.shutdown();
+    return new ABFSContentSummary(totalBytes.get(), directoryCount.get(),
+        fileCount.get(), totalBytes.get());
+  }
+
+  private void processDirectoryTree(Path path)
+      throws IOException, InterruptedException {
+    FileStatus[] fileStatuses = abfsStore.listStatus(path);
+
+    for (FileStatus fileStatus : fileStatuses) {
+      if (fileStatus.isDirectory()) {
+        queue.put(fileStatus);
+        processDirectory();
+        synchronized (this) {
+          if (!queue.isEmpty() && NUM_TASKS.get() < NUM_THREADS) {
+            NUM_TASKS.incrementAndGet();
+            executorService.submit(() -> {
+              try {
+                FileStatus fileStatus1;
+                while ((fileStatus1 = queue.poll(100, TimeUnit.MILLISECONDS))
+                    != null) {
+                  processDirectoryTree(fileStatus1.getPath());
+                }
+                NUM_TASKS.decrementAndGet();
```

Comment:
is a better place for this at the end of this method? say line 98

Reply:
yes, moved to finally